substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java

```diff
@@ -775,8 +775,19 @@ protected static void ensureDirectoryExists(Path dir) {
     private void prepareImageBuildArgs() {
         config.getBuilderJavaArgs().forEach(this::addImageBuilderJavaArgs);
         addImageBuilderJavaArgs("-Xss10m");
-        addImageBuilderJavaArgs(oXms + getXmsValue());
-        addImageBuilderJavaArgs(oXmx + getXmxValue(1));
+        String customSetXmx = customImageBuilderArgs.stream().filter(s -> s.startsWith(oXmx)).findFirst().orElse(null);
```

Comment:
You don't need to massage -Xmx/-Xms specific customImageBuilderArgs here. This will be done later if you move the
`/* Perform JavaArgs consolidation - take the maximum of -Xmx, minimum of -Xms */` block after
`addImageBuilderJavaArgs(customJavaArgs.toArray(new String[0]));` in `com.oracle.svm.driver.NativeImage#completeImageBuild`

Reply:
I think this is all you need in `NativeImage.java`:
```patch
diff --git a/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java b/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java
--- a/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java	(revision Staged)
+++ b/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java	(date 1618412218077)
@@ -806,7 +806,10 @@
         config.getBuilderJavaArgs().forEach(this::addImageBuilderJavaArgs);
         addImageBuilderJavaArgs("-Xss10m");
         addImageBuilderJavaArgs(oXms + getXmsValue());
-        addImageBuilderJavaArgs(oXmx + getXmxValue(1));
+        String xmxValue = getXmxValue(1);
+        if (!"0".equals(xmxValue)) {
+            addImageBuilderJavaArgs(oXmx + xmxValue);
+        }
         addImageBuilderJavaArgs("-Duser.country=US", "-Duser.language=en");
         /* Prevent JVM that runs the image builder to steal focus */
         if (OS.getCurrent() != OS.WINDOWS || JavaVersionUtil.JAVA_SPEC > 8) {
@@ -1178,13 +1181,6 @@
         }
         imageClasspath.addAll(customImageClasspath);
 
-        /* Perform JavaArgs consolidation - take the maximum of -Xmx, minimum of -Xms */
-        Long xmxValue = consolidateArgs(imageBuilderJavaArgs, oXmx, SubstrateOptionsParser::parseLong, String::valueOf, () -> 0L, Math::max);
-        Long xmsValue = consolidateArgs(imageBuilderJavaArgs, oXms, SubstrateOptionsParser::parseLong, String::valueOf, () -> SubstrateOptionsParser.parseLong(getXmsValue()), Math::max);
-        if (Long.compareUnsigned(xmsValue, xmxValue) > 0) {
-            replaceArg(imageBuilderJavaArgs, oXms, Long.toUnsignedString(xmxValue));
-        }
-
         imageBuilderJavaArgs.add("-Djdk.internal.lambda.disableEagerInitialization=true");
         // The following two are for backwards compatibility reasons. They should be removed.
         imageBuilderJavaArgs.add("-Djdk.internal.lambda.eagerlyInitialize=false");
@@ -1208,6 +1204,15 @@
         }
         addImageBuilderJavaArgs(customJavaArgs.toArray(new String[0]));
 
+        /* Perform JavaArgs consolidation - take the maximum of -Xmx, minimum of -Xms */
+        Long xmsValue = consolidateArgs(imageBuilderJavaArgs, oXms, SubstrateOptionsParser::parseLong, String::valueOf, () -> SubstrateOptionsParser.parseLong(getXmsValue()), Math::max);
+        Long xmxValue = consolidateArgs(imageBuilderJavaArgs, oXmx, SubstrateOptionsParser::parseLong, String::valueOf, () -> 0L, Math::max);
+        if (xmxValue != null) {
+            if (Long.compareUnsigned(xmsValue, xmxValue) > 0) {
+                replaceArg(imageBuilderJavaArgs, oXms, Long.toUnsignedString(xmxValue));
+            }
+        }
+
         /* Perform option consolidation of imageBuilderArgs */
 
         imageBuilderJavaArgs.addAll(getAgentArguments());
```

Reply:
Fixed as suggested.